server/src/main/java/io/pravega/schemaregistry/service/SchemaRegistryService.java

```diff
@@ -267,22 +269,25 @@ public SchemaRegistryService(SchemaStore store, ScheduledExecutorService executo
-        return RETRY.runAsync(() -> store.getGroupEtag(namespace, group)
-                                         .thenCompose(etag ->
-                                                 store.getGroupProperties(namespace, group)
-                                                      .thenCompose(prop -> {
-                                                          return Futures.exceptionallyComposeExpecting(store.getSchemaVersion(namespace, group, schema),
-                                                                  e -> Exceptions.unwrap(e) instanceof StoreExceptions.DataNotFoundException,
-                                                                  () -> { // Schema doesnt exist. Validate and add it
-                                                                      return validateSchema(namespace, group, schema, prop.getCompatibility())
-                                                                              .thenCompose(valid -> {
-                                                                                  if (!valid) {
-                                                                                      throw new IncompatibleSchemaException(String.format("%s is incompatible", schema.getType()));
-                                                                                  }
-                                                                                  return store.addSchema(namespace, group, schema, prop, etag);
-                                                                              });
-                                                                  });
-                                                      })), executor)
+        return RETRY.runAsync(() ->
+                store.getGroupEtag(namespace, group)
+                     .thenCompose(etag ->
+                             store.getGroupProperties(namespace, group)
+                                  .thenCompose(prop -> {
+                                      return Futures.exceptionallyComposeExpecting(store.getSchemaVersion(namespace, group, schema, getFingerprint(schema)),
+                                              e -> Exceptions.unwrap(e) instanceof StoreExceptions.DataNotFoundException,
+                                              () -> { // Schema doesnt exist. Validate and add it
+                                                  return validateSchema(namespace, group, schema, prop.getCompatibility())
+                                                          .thenCompose(valid -> {
+                                                              if (!valid) {
+                                                                  throw new IncompatibleSchemaException(String.format("%s is incompatible", schema.getType()));
+                                                              }
+                                                              // we will compute the fingerprint from normalized form.
```

Comment:
It sounds right to compute the fingerprint from the normalized form, but we can pass the orinal form and normalize it as part of the fingerprint call. I'm seeing the normalized form being passed around, and I'm not sure why we need to do it. I thought the normalized form was only necessary to compute the fingerprint, and nothing else. What am I missing?

Reply:
normalized form is only necessary for computing the fingerprint. 
and original form is what we use for storage.
for any other computation we could use either form as they are conceptually same and represent the same structure. 
the existing code was written with normalized form everywhere and there is no need to change it. we only change it at the time when we store it.

Reply:
The reason I'm raising this is that I don't think this call `SchemaRegistryService#addSchema` needs to be aware of normalization, and the additional parameter to pass the normalized version of the schema is unnecessary. Normalization can be hidden in the logic of `SchemaStore`.

Reply:
its actually the other way round -- the store layer doesnt care for normalization because it doesnt have the wherewithal to parse the schema. 
the service layer is capable of parsing the schema, validating that its well formed, and performs all compatibility checks on it.
So the service layer is capable of creating the normalized form (as being able to parse the schema is a prerequisite for it). 
the purpose of normalized form is so that we can compare whether two schemas were identical logically irrespective of whether the supplied schema string had different order of fields or extra whitespace or linefeed like characters. 
And since the intent here is to no longer store the normalized form, so it computes the hash of the normalized form and passes the hash with the original schema to the storage layer to store.

Reply:
I think my comment above should resolve this one, I'm finding this flow a bit unintuitive.

Reply:
> Normalization can be hidden in the logic of SchemaStore

As i mentioned, the abstraction is that schema store is simply a store with no capability of parsing the schema (which is a prerequisite for normalizing it). 
the service layer is responsible for normalizing and computing the hash. 

We also want to store the normalized form in the global pool of schemas 

note: the global table and group table is storage layer's implementation detail, but the concept of being able to look up groups where a specific schema is added is an api on the store's interface.

Reply:
We have discussed this offline, what's is that we agreed upon for this thread?